### PR TITLE
feat: adding new TrackSummary analysis

### PR DESCRIPTION
**Pull request by asalzburger**
This PR adds a new analysis application for track summary and perigee parameters analysis. It uses the newly available `TTree` from the `RootTrackSummaryPerformanceWriter`.

It also renames the two executables that are build from `ActsExample...` to `ActsAnalysis...` and bundles this with a renamed build flag `ACTS_BUILD_ANALYSIS_APPS`.

The built example is using a compilable ROOT script, and builds an executable with `boost::program_options` to steer the behaviour of the program. The program is executed in three stages:

**1. Handle creation and histogram booking with automated range detection:**

```shell
*** ACTS Perigee parameters and Track summary plotting
*** Adding file: tracksummary_full.root
*** Handle Preparation:

0%   10   20   30   40   50   60   70   80   90   100%
|----|----|----|----|----|----|----|----|----|----|
***************************************************
```

In this step, the handles (for residual, pulls, statistics) are prepared (for overall histograms and their binned versions), i.e. the residual accessors and error accessors are created and for them and the cut structs are created. The range histograms are created with `TChain::Draw(...)` which can take a long time for a large number of entries, because this `Draw(...)` command will be called for every single step. To shorten the time, one can regulate the number of events for peaking out the range with the `--peak-events` options. Additionally, one can write out the estimated ranges into a `JSON` config file using the `--config-output` option. The config file can in a next plotting iteration be loaded and the time consuming range-estimation would be saved.

**2.  Loop over the entries and filling of the overall/binned histograms**

In a second step, the loop over all events (or, if required, over a subset of events clarified with the `--events,-n` option) and the all the handles are addressed. As every single handle already has the correct value accessor and cut struct, the associated histograms will be filled with the correct values, but only when the cut struct accept the track in question.

```shell
*** Event Loop:

0%   10   20   30   40   50   60   70   80   90   100%
|----|----|----|----|----|----|----|----|----|----|
***************************************************
```

**3. Bin analysis and writing** 

In the final step, the histograms (all/within the bin matrix) are analysed and global 2D histograms created. Out of the 2D histograms the (bin-wise and integral) projection histograms (called `_pX` and `_pY`) are created.

```shell
*** Bin/Projection Analysis:

0%   10   20   30   40   50   60   70   80   90   100%
|----|----|----|----|----|----|----|----|----|----|
***************************************************

0%   10   20   30   40   50   60   70   80   90   100%
|----|----|----|----|----|----|----|----|----|----|
***************************************************
```

The full set of commands to steer the program are as follows:
```shell
*** ACTS Perigee parameters and Track summary plotting
*** Usage::
  -h [ --help ]                         Display this help message
  -s [ --silent ]                       Silent mode (without X-window/display).
  -n [ --events ] arg (=0)              (Optionally) limit number of events to
                                        be processed.
  -p [ --peak-events ] arg (=0)         (Optionally) limit number of events for
                                        the range peaking.
  -i [ --input ] arg                    Input ROOT file(s) containing the input
                                        TTree.
  -t [ --tree ] arg (=tracksummary)     Input TTree/TChain name.
  -o [ --output ] arg                   Output ROOT file with histograms
  --hist-bins arg (=61)                 Numer of bins for the residual/pull
                                        histograms
  --pull-range arg (=5)                 Number of sigmas for the pull range.
  --eta-bins arg (=10)                  Number of bins in eta.
  --eta-range MIN:MAX (=-3:3)           Range for the eta bins.
  --phi-bins arg (=10)                  Number of bins in phi.
  --phi-range MIN:MAX (=-3.14159:3.14159)
                                        Range for the phi bins.
  --pt-borders arg                      Transverse momentum borders.
  --config-output arg                   Output histrogram configuration json
                                        file.
  --config-input arg                    Input histrogram configuration json
                                        file.
  --all                                 Process all residual/pull and auxiliary
                                        parameters
  --d0                                  Residual/pulls for d0
  --z0                                  Residual/pulls for z0
  --phi0                                Residual/pulls for phi0
  --theta0                              Residual/pulls for theta0
  --qop                                 Residual/pulls for qop
  --time                                Residual/pulls for time
  --pt                                  Residual/pulls for pt
  --chi2ndf                             Auxiliary information for chi2ndf
  --measurements                        Auxiliary information for measurements
  --holes                               Auxiliary information for holes
  --outliers                            Auxiliary information for outliers
  --shared                              Auxiliary information for shared
```

One can chose to either process all parameters (`--all`) or only individual ones (e.g. `--d0`).

A typical command for a first pass analysis would be:

```shell
ActsAnalysisTrackSummary  -i <input_file> -t tracksummary -o <output_file> --hist-bins 61 --phi-bins 36 --eta-bins 20 --pt-borders 0:1:5:10:25:50:75:100 --config-output <configuration_file> -s --peak-events 200 -n1000 --all
```

This will take 200 peak events (let's assume one event has 100 tracks, i.e. 20000 tracks) for the histogram range and then process 100k tracks into the bins). The statistics would not be very good, so one would probably use either more peak events or use less bins.

In a repeated analysis, one could run over the full stats (practically only event loop and analysis) by re-using the range `JSON` configuration as:

```shell
ActsAnalysisTrackSummary -i <input_file> -t tracksummary -o <output_file> --hist-bins 61 --phi-bins 36 --eta-bins 20 --pt-borders 0:1:5:10:25:50:75:100 --config-input <configuration_file>  -s --all
```

This will now process all entries in the `TChain` for all parameters, once can change the number of histogram bins, but if either `phi`, `eta` binning or the `pt-borders` sequence is changed, the ranges will be re-evaluated.

The histograms are then written into the `<output_file>` file in this example.

There are several classes of histograms being written out.

**Overall plots:**

These are named `res_d0_all`, `pull_qp_all` for residual/pull histograms or `chi2ndf_all` for auxiliary histograms:

Screenshot 2021-09-23 at 10 29 24

Screenshot 2021-09-23 at 10 29 46

**Binned plots:**

All histograms also exist for the chosen `[eta, phi]` and `[eta, pt]` binning, they would then be called, e.g. `res_d0_eta3_pt4` etc.

Screenshot 2021-09-23 at 09 54 10

**2D summary plots:**

Given the binning 2D summary plots exist for the chosen parameters, these are e.g.
`res_rms_d0_eta_phi` and `res_rms_d0_eta_pt` and so on, the summary plots exist for residual parameters as:

 * RMS or residual distribution: `res_rms_<parameter>_<binning>`
 * mean of residual distribution: `res_mean_<parameter>_<binning>`
 * sigma of gaussian fit to pull distribution: `pull_sigma_<parameter>_<binning>`
 * mean of gaussian fit to pull distribution: `pull_mean_<parameter>_<binning>`

Screenshot 2021-09-23 at 09 50 05

For auxiliary plots these exist as well, but simply as mean with associated error, e.g `chi2ndf_<binning>`

Screenshot 2021-09-23 at 09 45 52

Screenshot 2021-09-23 at 09 44 46

**Projection plots:**

For convenience, also the `ProjectionX()` and `ProjectionY()` histograms integral form an for the relevant bin rows are also created.

### CODE REVIEW

**Comment by asalzburger** [comment]
Thinking about it, I could add a `StatisticsHandle` for hits, holes and outliers, and process them in the same manner, then I *think* we should call this `ActsAnalysisTrackSummary`.

What do you think @XiaocongAi ?

I also think we should move the `BoundParameters` estimation not he surfaces to a similar setup, because it would allow us to do robust RMS/gaus fitting, this could be a plug-in into the analysis.

**Comment by XiaocongAi** [comment]
Hi @asalzburger ,  thank you for the detailed documentation. I will play it around following your instructions and review the PR a bit later.

**Comment by XiaocongAi** [comment]
> Thinking about it, I could add a `StatisticsHandle` for hits, holes and outliers, and process them in the same manner, then I _think_ we should call this `ActsAnalysisTrackSummary`.
> 
> What do you think @XiaocongAi ?
> 
> I also think we should move the `BoundParameters` estimation not he surfaces to a similar setup, because it would allow us to do robust RMS/gaus fitting, this could be a plug-in into the analysis.

I think that's a good idea. It can helps us check the holes and outliers etc. immediately. And yes, we should do the same with the bound parameters.

**Comment by asalzburger** [comment]
Hi, so that's my shot at a systematic track summary analysis @XiaocongAi can you give this a try if it works for your?

 @plariono might also be interested in this.

**Comment by asalzburger** [comment]
Hi @XiaocongAi - I've added a protection against failing fits, probably this would help you.

File: Examples/Scripts/TrackingPerformance/trackSummaryAnalysis.C [hunk 1]
```diff
@@ -0,0 +1,949 @@
...
+#include <TVectorF.h>
+
+#include "CommonUtils.h"
+#include "TreeReader.h"
+
+using namespace ROOT;
+
+/// This ROOT script will plot the residual and pull of perigee track parameters
+/// (d0, z0, phi, theta, q/p, pT, t) from root file produced by the
+/// TrackFitterPerformanceWriter
```
Comment (XiaocongAi) [inline Examples/Scripts/TrackingPerformance/trackSummaryAnalysis.C:37]:
```suggestion
/// RootTrajectorySummaryWriter
```

File: Examples/Scripts/TrackingPerformance/TrackSummary.cpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+       value<Interval>()->value_name("MIN:MAX")->default_value({-3.0, 3.0}),
+       "Range for the eta bins.");
+    ao("phi-bins", value<unsigned int>()->default_value(10),
+       "Number of bins in phi.");
+    ao("phi-range",
+       value<Interval>()->value_name("MIN:MAX")->default_value({-M_PI, M_PI}),
+       "Range for the phi bins.");
+    ao("pt-borders", value<VariableReals>(), "Transverse momentum borders.");
+    ao("config-output", value<std::string>()->default_value(""),
+       "Output histrogram configuration json file.");
```
Comment (XiaocongAi) [inline Examples/Scripts/TrackingPerformance/TrackSummary.cpp:68]:
```suggestion
       "(Optional) output histrogram configuration json file.");
```

File: Examples/Scripts/TrackingPerformance/TrackSummary.cpp [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    ao("config-input", value<std::string>()->default_value(""),
+       "Input histrogram configuration json file.");
```
Comment (XiaocongAi) [inline Examples/Scripts/TrackingPerformance/TrackSummary.cpp:70]:
```suggestion
       "(Optional) input histrogram configuration json file.");
```

**Review by XiaocongAi — APPROVED** [review]
Hi @asalzburger , it works well now. Thank you for the nice scripts, which makes our life much easier.

**Comment by asalzburger** [comment]
This was approved already, merging it now.

**Pull request merged**